aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -82,37 +74,89 @@ struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct
+    char* return_val;
+    uint8_t index;
+     uint8_t entry_counter;
+     struct aesd_buffer_entry* entryptr = NULL;
+    //struct aesd_buffer_entry* aesd_buff_entry = (struct aesd_buffer_entry*) malloc(sizeof(struct aesd_buffer_entry));
     
-    memcpy(aesd_buff_entry, add_entry, sizeof(*add_entry));
+    //memcpy(aesd_buff_entry, add_entry, sizeof(*add_entry));
 
     if(buffer == NULL)
     {
-        return;
+        return NULL;
+    }
+        if(buffer->full == true)
+    {
+        return_val = (char*)buffer->entry[buffer->in_offs].buffptr;
     }
-        
-    buffer->entry[buffer->in_offs].buffptr = aesd_buff_entry->buffptr;
-    buffer->entry[buffer->in_offs].size = aesd_buff_entry->size;
+    PDEBUG("|-|writing %s ", add_entry->buffptr);    
+    PDEBUG("add entry ptr %p ", add_entry->buffptr);   
+     
+    buffer->entry[buffer->in_offs] = *add_entry;
+    PDEBUG("circ buff entry buffptr %p size %ld", buffer->entry[buffer->in_offs].buffptr, sizeof(buffer->entry[buffer->in_offs].buffptr));  
+    // memcpy((char*)buffer->entry[buffer->in_offs].buffptr, add_entry->buffptr, add_entry->size);
     if(buffer->full == true)
     {
         buffer->out_offs = (buffer->out_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
     }
+    else
+    {
+        return_val = NULL;
+    }
     
     buffer->in_offs = (buffer->in_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
 
-    buffer->full = (buffer->in_offs == buffer->out_offs);
-    LOG_DBG("out %d in %d full %d", buffer->out_offs, buffer->in_offs, buffer->full);
+
+        PDEBUG("********************************************************************");
+
+    for(index=buffer->out_offs, entryptr=&((buffer)->entry[index]), entry_counter = 0; \
+			entry_counter < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; \
+			entry_counter++, index = (index + 1)%AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED , entryptr=&((buffer)->entry[index]))
+    {
+        PDEBUG("id <%d>, buf %s, size: %ld", index, (buffer)->entry[index].buffptr, entryptr->size);
+    }
+        PDEBUG("********************************************************************");
+
+    
+    if( buffer->in_offs == buffer->out_offs)
+       buffer->full = true;
+       else
+       buffer->full = false;
+    PDEBUG("out %d in %d full %d", buffer->out_offs, buffer->in_offs, buffer->full);
+
+    return return_val;    
 }
 
 /**
 * Initializes the circular buffer described by @param buffer to an empty struct
 */
 void aesd_circular_buffer_init(struct aesd_circular_buffer *buffer)
 {
+     PDEBUG("inside cicrcular buffer init");
     memset(buffer,0,sizeof(struct aesd_circular_buffer));
 }
 
+void aesd_cicular_buffer_free(struct aesd_circular_buffer *buffer)
+{
+    uint8_t index;
+    struct aesd_buffer_entry *entry;
+
+    AESD_CIRCULAR_BUFFER_FOREACH(entry, buffer, index)
+    {
+
+        if (entry->buffptr != NULL)
+        {
 
+#ifdef __KERNEL__
+            kfree(entry->buffptr);
```

Comment:
@rajatchaple I'd put the logic to free outside this file and in the individual application.   That way if the application doesn't dynamically allocate the buffer it won't try to free here.